dotnet/src/Extensions/Reliability.Polly/Config/DefaultHttpRetryHandler.cs

```diff
@@ -0,0 +1,135 @@
+
+        policyBuilder = policyBuilder.Or<Exception>(e => this._config.RetryableExceptionTypes.Contains(e.GetType()));
+
+        var timeoutPolicy = Policy.TimeoutAsync<HttpResponseMessage>(this._config.MaxTotalRetryTime);
+
+        var retryPolicy = policyBuilder.WaitAndRetryAsync(
+            retryCount: this._config.MaxRetryCount,
+            sleepDurationProvider: (retryCount, context, timeSpan) =>
+            {
+                return this.GetWaitTime(retryCount, context.Result);
+            },
+            (outcome, timespan, retryCount, _) =>
+            {
+                this._logger?.LogWarning(
+                    "Error executing action [attempt {CurrentRetryAttempt} of {MaxRetryAttempts}]. Reason: {StatusCode}. Will retry after {WaitMilliseconds}ms.",
+                    retryCount,
+                    this._config.MaxRetryCount,
+                    outcome.Result.StatusCode,
+                    timespan.TotalMilliseconds);
+                return Task.CompletedTask;
+            });
+
+        return Policy.WrapAsync<HttpResponseMessage>(timeoutPolicy, retryPolicy);
+    }
+
+    private bool IsRetryDisabled()
+    {
+        return this._config.MaxRetryCount == 0;
+    }
+
+    /// <summary>
+    /// Get the wait time for the next retry.
+    /// </summary>
+    /// <param name="retryCount">Current retry count</param>
+    /// <param name="response">The response message that potentially contains RetryAfter header.</param>
+    private TimeSpan GetWaitTime(int retryCount, HttpResponseMessage? response)
+    {
+        // If the response contains a RetryAfter header, use that value
+        // Otherwise, use the configured min retry delay
+        var retryAfter = response?.Headers.RetryAfter?.Date.HasValue == true
+            ? response?.Headers.RetryAfter?.Date - this._timeProvider.GetCurrentTime()
+            : (response?.Headers.RetryAfter?.Delta) ?? this._config.MinRetryDelay;
+        retryAfter ??= this._config.MinRetryDelay;
+
+        // If the retry delay is longer than the max retry delay, use the max retry delay
+        var timeToWait = retryAfter > this._config.MaxRetryDelay
+            ? this._config.MaxRetryDelay
+            : retryAfter < this._config.MinRetryDelay
+                ? this._config.MinRetryDelay
+                : retryAfter ?? default;
+
+        // If exponential backoff is enabled, double the delay for each retry
+        if (this._config.UseExponentialBackoff)
+        {
+            timeToWait = TimeSpan.FromTicks(timeToWait.Ticks * 2);
+        }
+
+        return timeToWait;
+    }
+
+    /// <summary>
+    /// Interface for a time provider, primarily to enable unit testing.
+    /// </summary>
+    internal interface ITimeProvider
+    {
+        DateTimeOffset GetCurrentTime();
+    }
+
+    internal sealed class DefaultTimeProvider : ITimeProvider
+    {
+        public DateTimeOffset GetCurrentTime()
+        {
+            return DateTimeOffset.UtcNow;
+        }
+    }
```

Comment:
Consider moving the TimeProvider interface and class to the abstractions project so it can be used by other components.

Reply:
I think this deserves a dedicated PR. As TimeProvider as an abstraction would only make sense if used by more than one component. 

I can be wrong but currently I don't see any other Kernel component that uses the Clock.

Reply:
These were added to improve testability.